draft-ietf-jsonpath-base.md

```diff
@@ -163,65 +163,61 @@ Importantly "object" and "array" in particular do not take on a
 generic meaning, such as they would in a general programming context.
 
 Member:
-: A name/value pair in a JSON object.  (Not itself a Value.)
+: A name/value pair in a JSON object.  (Not itself a value.)
 
 Name:
 : The name in a name/value pair constituting a member.  (Also known as
   "key", "tag", or "label".)
+  This is also used in {{-json}}, but that specification does not 
+  formally define it.
+  It is included here for completeness.
 
 Element:
-: A Value in an array.  (Also used with a distinct meaning in XML
+: A value in an array.  (Also used with a distinct meaning in XML
   context for XML elements.)
 
 Index:
 : A non-negative integer that identifies a specific element in an array.
-  The term, in particular in the form indexing, is also sometimes used
-  for either an index or a member name, when both arrays and objects
-  can be fed to the indexing operation.
 
 Query:
 : Short name for JSONPath expression.
 
 Argument:
-: Short name for the Value a JSONPath expression is applied to.
+: Short name for the value a JSONPath expression is applied to.
 
 Node:
-: The pair of a Value along with its location within the Argument.
+: The pair of a value along with its location within the argument.
 
 Root Node:
-: The unique Node whose Value is the entire Argument.
+: The unique node whose value is the entire argument.
 
 Nodelist:
-: Output of applying a query to an Argument, manifested as a list of Nodes.
+: A list of nodes.  <!-- ordered list?  Maybe TBD by issues #27 and #60 -->
+  The output of applying a query to an argument is manifested as a list of nodes.
   While this list can be represented in JSON, e.g. as an array, the
   Nodelist is an abstract concept unrelated to JSON values.
 
 <!--
   I feel like the below definition should be for "normalized path,"
```

Comment:
The definitional aspect of the Output Path is that it is output.  The fact that it is normalized is certainly interesting, but not definitional.  Normalized paths could occur anywhere.  Output Paths occur at the output.

Reply:
- `Node` is defined as a value and a location.
- `Nodelist` is defined as a collection of nodes.  The definition also mentions that the output of a query is a nodelist.
- Both the "Processing Model" and "Sematics" sections also mention that the output is a nodelist.  Moreover, "Processing Model" says that each selector produces a nodelist which is then used as the input to the next selector.

It seems to me that the output is already well-defined by the above.

If we also declare that the location portion of a node needs to be represented with a normalized path, then we get for free that the paths in the output are a normalized path (because the output is a list of nodes), and we don't need to define what an "output path" is.

However, "normalized path" still needs to be defined.  This is illustrated by your statement:

> Normalized paths could occur anywhere.

I don't think we need to have a special-case definition for a path that occurs in the output.

Reply:
I think we agree with the following ...

* JSONPath is *selecting* only, not *processing*.
* Analytical post processing on the query result is heavily demanded by users.

I do like that brillant *extension points* concept a lot. So question is: Can we open the door for analytical post processing via those *extension points*? If so, the spec must provide certainty to implementors regarding *output path* format, which should be either/or ...

* list of selected values
* list of normalized pathes

So an *output path* always should have *normalized path* format in my opinion.

Reply:
> So an output path always should have normalized path format in my opinion.

I'm not arguing this.  I'm questioning whether we need "output path" as an explicit definition.

To me, "output path" is just a normalized path that happens to be in the output.  There's nothing really special about it that requires a definition.  We _do_ need to define the output somewhere, but it's not a definition or term; it's a feature.

"Normalized path," on the other hand, is precisely what the current definition is describing:

> A simple form of JSONPath expression that identifies a node by providing a query that results in exactly that node.